pkg/controller/grouping/group_entity_index.go

```diff
@@ -0,0 +1,719 @@
+}
+
+func (i *groupEntityIndex) match(entityType entityType, label labels.Set, namespace string, sel *types.GroupSelector) bool {
+	objSelector := sel.PodSelector
+	if entityType == externalEntityType {
+		objSelector = sel.ExternalEntitySelector
+	}
+	if sel.Namespace != "" {
+		if sel.Namespace != namespace {
+			// Pods or ExternalEntities must be matched within the same Namespace.
+			return false
+		}
+		if objSelector != nil && !objSelector.Matches(label) {
+			// podSelector or externalEntitySelector exists but doesn't match the ExternalEntity or Pod's labels.
+			return false
+		}
+		return true
+	}
+	if sel.NamespaceSelector != nil {
+		if !sel.NamespaceSelector.Empty() {
+			namespaceLabels, exists := i.namespaceLabels[namespace]
+			if !exists {
+				return false
+			}
+			if !sel.NamespaceSelector.Matches(namespaceLabels) {
+				// Pod's Namespace do not match namespaceSelector.
+				return false
+			}
+		}
+		if objSelector != nil && !objSelector.Matches(label) {
+			// ExternalEntity or Pod's Namespace matches namespaceSelector but
+			// labels do not match the podSelector or externalEntitySelector.
+			return false
+		}
+		return true
+	}
+	if objSelector != nil {
+		// Selector only has a PodSelector/ExternalEntitySelector and no sel.Namespace.
+		// Pods/ExternalEntities must be matched from all Namespaces.
+		if !objSelector.Matches(label) {
+			// pod/ee labels do not match PodSelector/ExternalEntitySelector.
+			return false
+		}
+		return true
+	}
+	// TODO: if none of the above selector is set, it means matching all entities or none?
+	// This is only possible for ClusterGroup. From previous code, it means matching none.
+	return false
+}
+
+func entityAttrsUpdated(oldEntity, newEntity metav1.Object) bool {
+	switch oldValue := oldEntity.(type) {
+	case *v1.Pod:
+		// For Pod, we only care about PodIP and NodeName update.
+		// Some other attributes we care are immutable, e.g. the named ContainerPort.
+		newValue := newEntity.(*v1.Pod)
+		if oldValue.Status.PodIP != newValue.Status.PodIP {
+			return true
+		}
+		if oldValue.Spec.NodeName != newValue.Spec.NodeName {
+			return true
+		}
+		return false
+	case *v1alpha2.ExternalEntity:
+		newValue := newEntity.(*v1alpha2.ExternalEntity)
+		if !reflect.DeepEqual(oldValue.Spec, newValue.Spec) {
+			return true
+		}
+		return false
+	}
+	return false
+}
+
+// getEntityKey returns the entity key used in entityItems.
+func getEntityKey(entityType entityType, entity metav1.Object) string {
+	return string(entityType) + "/" + entity.GetNamespace() + "/" + entity.GetName()
+}
+
+// getEntityKeyByName returns the entity key used in entityItems.
+func getEntityKeyByName(entityType entityType, namespace, name string) string {
+	return string(entityType) + "/" + namespace + "/" + name
+}
+
+// getLabelKey returns the label key used in labelItems.
+func getLabelKey(entityType entityType, obj metav1.Object) string {
```

Comment:
I find the term "label key" a bit confusing because in K8s a "label key" already means something (the "name" of the label).

Reply:
updated to `getLabelItemKey`, so do others.